clang/lib/Driver/Driver.cpp

```diff
@@ -3858,13 +3870,16 @@ class OffloadingActionBuilder final {
         Action *A(HostAction);
         // Only check for FPGA device information when using fpga SubArch.
         if (Args.hasArg(options::OPT_fintelfpga) &&
-            HostAction->getType() != types::TY_FPGA_AOCR &&
-            HostAction->getType() != types::TY_FPGA_AOCX &&
             !(HostAction->getType() == types::TY_Object &&
               isObjectFile(InputName))) {
-          if (HasFPGADeviceBinary(C, InputArg->getAsString(Args), true))
+          // Type FPGA aoco is a special case for -foffload-static-lib.
+          if (HostAction->getType() == types::TY_FPGA_AOCO) {
+            if (!hasFPGABinary(C, InputName, types::TY_FPGA_AOCO))
+              return false;
+            A = C.MakeAction<InputAction>(*InputArg, types::TY_FPGA_AOCO);
+          } else if (hasFPGABinary(C, InputName, types::TY_FPGA_AOCX))
```

Comment:
Could it be the case that both `aocx` and `aoco` objects are passed, e.g. for linking?

Reply:
You can pass both aocx and aoco, but they are handled differently.  aoco will only be unbundled from the -foffload-static-lib option and aocx is unbundled earlier when regular objects are unbundled.